interfaces/apparmor/spec.go

```diff
@@ -146,6 +166,70 @@ func (spec *Specification) AddSnippet(snippet string) {
 	}
 }
 
+// RegisterSnippetKey adds a key to the list of valid keys
+func (spec *Specification) RegisterSnippetKey(key interfaces.SnippetKey) {
+	if spec.snippetKeys == nil {
+		spec.snippetKeys = make(map[interfaces.SnippetKey]bool)
+	}
+	if _, ok := spec.snippetKeys[key]; ok {
+		logger.Panicf("priority key %s is already registered", key.String())
+	}
+	spec.snippetKeys[key] = true
+}
+
+// AddPrioritizedSnippet adds a new apparmor snippet to all applications and hooks using the interface,
+// but identified with a key and a priority. If no other snippet exists with that key, the snippet is
+// added like with AddSnippet, but if there is already another snippet with that key, the priority of
+// both will be taken into account to decide whether the new snippet replaces the old one, is appended
+// to it, or is just ignored. The key must have been previously registered using RegisterSnippetKey().
+func (spec *Specification) AddPrioritizedSnippet(snippet string, key interfaces.SnippetKey, priority uint) {
+	if _, ok := spec.snippetKeys[key]; !ok {
+		logger.Panicf("priority key %s is not registered", key.String())
```

Comment:
Quick question to @pedronis if we should use it as-is (panic).

My personal feeling - yes we should and we should make sure that all backends can recover from panicking interface code and fail gracefully (without taking snapd down).

Reply:
I fully agree, and in fact I was expecting a comment about this to decide how to proceed if, due to a coding error (because it can happen only due to coding error, a typo for example), this happen. Should it just show a warning/error and don't add the snippet? Should it fully disconnect that plug?

It is supposed that this kind of error should (or I even would say that it must) be detected during debugging, in the sense that not detecting this means that the tests are incomplete because not all the possible running paths are being tested...

Reply:
@zyga @pedronis I did a change in the code: now you have to ask to your Spec instance for a specific SnippetKey, and it will return it and tell if it is registered or not, so you can decide what to do in case of error. The panic still happens if you try to use a SnippetKey that your Spec instance told you that isn't registered, so now it can be done safe. It also makes SnippetKey a true opaque type. What do you think?